### PR TITLE
Added nodes for managing timers

### PR DESCRIPTION
**Pull request by davidmarr**
Added nodes for managing timers (TimerNodes.lua)

### CODE REVIEW

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua [hunk 1]
```diff
@@ -0,0 +1,274 @@
+local Timer = require("Engine.Timer")
+LevelFuncs.noA = function() end
```
Comment (Lwmte) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: What is this? If you have to pass null function, you can try to pass `nil` instead.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+-- !Name "Create Timer with Function"
+-- !Conditional "False"
+-- !Description "after a specified number of seconds, the specified thing happens"
```
Comment (Lwmte) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: Please capitalize the sentence.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua [hunk 3]
```diff
@@ -0,0 +1,274 @@
+local Timer = require("Engine.Timer")
+LevelFuncs.noA = function() end
+
+-- !Name "Create Timer with Function"
```
Comment (Lwmte) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: Function and Timer words should not be capitalized. Stick to this rule onwards please.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+-- !Conditional "False"
+-- !Description "after a specified number of seconds, the specified thing happens"
+-- !Section "Timer"
+-- !Arguments "NewLine, String, 57,[ NoMultiline ], Timer name"
```
Comment (Lwmte) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: Missing space between comma and [

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+-- !Arguments "Numerical, 30, [ 0 | 1000 | 2 ], The duration of the timer in seconds"
+-- !Arguments "Boolean , 13, Loop"
+-- !Arguments "NewLine, Boolean , 33, Show Minutes" "Boolean , 33, Show Seconds" "Boolean , 33, Show Deciseconds"
```
Comment (Lwmte) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua:10]: Remove excessive capitalizing, as stated above.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+-- !Arguments "NewLine, LuaScript, The function to call when the time is up"
+LevelFuncs.Engine.Node.CreateTimerWithFunction = function(name, time, loop, minutes, seconds, deciseconds, fucs)
```
Comment (Lwmte) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: `fucs`?

**Review by Lwmte — COMMENTED** [review]
No notes from my side except formatting. Should be tested by someone (@Kubsy ?)

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua (hunk 1 above)
Comment (davidmarr) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: image
image
If I try to pass a null value there is an error

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua (hunk 6 above)
Comment (davidmarr) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: I renamed the **_fucs_** parameter to **_luaFunction_**

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua (hunk 1 above)
Comment (Lwmte) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: You need to fix underlying code then. It seems @hispidence forgot to add a nil check to a callback function.

Please go to `Scripts\Engine\Timer.lua` file and go to line 122. You will find this block of code:

```
if t.remainingTime <= 0 then
	if not t.loop then
		t.active = false
	else
		t.remainingTime = t.remainingTime + t.totalTime
	end
	t.func(table.unpack(t.funcArgs))
end
```

Change it to:

```
if t.remainingTime <= 0 then
	if not t.loop then
		t.active = false
	else
		t.remainingTime = t.remainingTime + t.totalTime
	end
	
	if (t.func ~= nil) then
		t.func(table.unpack(t.funcArgs))
	end
end
```

This should fix your issue. If it does, I will push this change to TEN repo.

Comment (davidmarr) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Timers.lua]: with this edit the nil value is accepted and of course the timer does not execute any function when finished. The other timers with an assigned function work perfectly!

**Review by Kubsy — APPROVED** [review]

**Pull request merged**